### ISSUE TITLE
beam example solution full of nans

### ISSUE THREAD

**Issue by whoburg** [hoburg/gpkit#770]
The solution `w_gp` to `docs/source/examples/beam.py` is full of nans. This does not seem right. Below is a copy of the example code with one line inserted to `print w_gp`:

```python
"""
A simple beam example with fixed geometry. Solves the discretized
Euler-Bernoulli beam equations for a constant distributed load
"""
import numpy as np
from gpkit import Variable, VectorVariable, Model, units
from gpkit.small_scripts import mag


class Beam(Model):
    """Discretization of the Euler beam equations for a distributed load.

    Arguments
    ---------
    N : int
        Number of finite elements that compose the beam.
    L : float
        [m] Length of beam.
    EI : float
        [N m^2] Elastic modulus times cross-section's area moment of inertia.
    q : float or N-vector of floats
        [N/m] Loading density: can be specified as constants or as an array.
    """
    def __init__(self, N=4, **kwargs):
        EI = Variable("EI", 1e4, "N*m^2")
        dx = Variable("dx", "m", "Length of an element")
        L = Variable("L", 5, "m", "Overall beam length")
        q = VectorVariable(N, "q", 100*np.ones(N), "N/m",
                           "Distributed load at each point")
        V = VectorVariable(N, "V", "N", "Internal shear")
        V_tip = Variable("V_{tip}", 0, "N", "Tip loading")
        M = VectorVariable(N, "M", "N*m", "Internal moment")
        M_tip = Variable("M_{tip}", 0, "N*m", "Tip moment")
        th = VectorVariable(N, "\\theta", "-", "Slope")
        th_base = Variable("\\theta_{base}", 0, "-", "Base angle")
        w = VectorVariable(N, "w", "m", "Displacement")
        w_base = Variable("w_{base}", 0, "m", "Base deflection")
        # below: trapezoidal integration to form a piecewise-linear
        #        approximation of loading, shear, and so on
        # shear and moment increase from tip to base (left > right)
        shear_eq = (V >= V.right + 0.5*dx*(q + q.right))
        shear_eq[-1] = (V[-1] >= V_tip)  # tip boundary condition
        moment_eq = (M >= M.right + 0.5*dx*(V + V.right))
        moment_eq[-1] = (M[-1] >= M_tip)
        # slope and displacement increase from base to tip (right > left)
        theta_eq = (th >= th.left + 0.5*dx*(M + M.left)/EI)
        theta_eq[0] = (th[0] >= th_base)  # base boundary condition
        displ_eq = (w >= w.left + 0.5*dx*(th + th.left))
        displ_eq[0] = (w[0] >= w_base)
        # minimize tip displacement (the last w)
        Model.__init__(self, w[-1],
                       [shear_eq, moment_eq, theta_eq, displ_eq,
                        L == (N-1)*dx], **kwargs)


b = Beam(N=6, substitutions={"L": 6, "EI": 1.1e4, "q": 110*np.ones(10)})
b.zero_lower_unbounded_variables()
sol = b.solve(verbosity=1)
w_gp = sol("w")  # deflection along beam

L, EI, q = sol("L"), sol("EI"), sol("q")
x = np.linspace(0, mag(L), len(q))*units.m  # position along beam
q = q[0]  # assume uniform loading for the check below
w_exact = q/(24.*EI) * x**2 * (x**2 - 4*L*x + 6*L**2)  # analytic soln

print w_gp
assert max(abs(w_gp - w_exact)) <= 1e-2*units.m
```

The printout is:
```
[  0.  nan  nan  nan  nan  nan] meter
```

**Comment by aechevarria1** [comment]
Gonna to take a look at this

**Comment by whoburg** [comment]
👍

**Comment by aechevarria1** [comment]
Okay so I found a potential issue. I was playing around with the boundary condition variables, and I found that when ```w_base``` is 0, it causes ```w_gp``` to be an array of ```nan```. However, if you make ```w_base``` non-zero, then ```w_gp``` outputs an array of numbers. So this input

``` python 
 V_tip = Variable("V_{tip}", 0, "N", "Tip loading")
M_tip = Variable("M_{tip}", 0, "N*m", "Tip moment")
th_base = Variable("\\theta_{base}", 0, "-", "Base angle")
w_base = Variable("w_{base}", 1, "m", "Base deflection")
```
produces this output

```[ 1.          1.106272    1.383616    1.75945601  2.18195201  2.62000002]```
So that may be an issue. I also noticed a couple other things. Whenever a boundary condition is substituted into one of the corresponding VectorVariables, if you print the Vector Variable, it doesn't show the updated value.

``` python
shear_eq = (V >= V.right + 0.5*dx*(q + q.right))
print shear_eq
shear_eq[-1] = (V[-1] >= V_tip)  # tip boundary condition
print shear_eq
print shear_eq[-1]
```

produces this output.

``` python
vec{V} >= [0.5*dx*q_(0,) + 0.5*dx*q_(1,) + V_(1,), 0.5*dx*q_(1,) + 0.5*dx*q_(2,) + V_(2,), 0.5*dx*q_(2,) + 0.5*dx*q_(3,) + V_(3,), 0.5*dx*q_(3,) + 0.5*dx*q_(4,) + V_(4,), 0.5*dx*q_(4,) + 0.5*dx*q_(5,) + V_(5,), 0.5*dx*q_(5,)]
vec{V} >= [0.5*dx*q_(0,) + 0.5*dx*q_(1,) + V_(1,), 0.5*dx*q_(1,) + 0.5*dx*q_(2,) + V_(2,), 0.5*dx*q_(2,) + 0.5*dx*q_(3,) + V_(3,), 0.5*dx*q_(3,) + 0.5*dx*q_(4,) + V_(4,), 0.5*dx*q_(4,) + 0.5*dx*q_(5,) + V_(5,), 0.5*dx*q_(5,)]
V_(5,) >= V_{tip}
```

so printing out the vector shows no change. But if you print out the specific index that was changed, it does show it the value was substituted.

The last thing I noticed is that if you assign a Variable a value of 0, then try to print out the value of the Variable, you get an error

``` python
z = Variable("z",0)
print z.value
```

outputs at the end of the traceback

``` python
/home/angel/.local/lib/python2.7/site-packages/gpkit/nomials/nomial_math.pyc in __init__(self, exps, cs, require_positive, simplify, **descr)
    110             from .. import SIGNOMIALS_ENABLED
    111             if require_positive and not SIGNOMIALS_ENABLED:
--> 112                 raise ValueError("each c must be positive.")
    113             self.__class__ = Signomial
    114         else:
ValueError: each c must be positive
```

so there's a signomial error. Was this intentional to prevent Variables from being assigned 0? Does the model solve properly without needing to access the value using that command?

**Comment by whoburg** [comment]
ok @aechevarria1 good sleuthing. All three of your observations are important for us to understand and fix.

I noticed a 4th issue / possible cause: in running `beam.py`, `sol("w")` and `sol["variables"]["w"]` both evaluate to `<Quantity([  0.  nan  nan  nan  nan  nan], 'meter')>`, whereas `sol["freevariables"]["w"]` evaluates to `<Quantity([        nan  0.10627203  0.38361602  0.75945602  1.18195202  1.62000002], 'meter')>`.

I believe this is caused by the line `result["variables"].update(result["constants"])` in `geometric_program.py`, where in this case some of the keys are VectorVariables that have some of their indices constant and some free.

@bqpd, I attempted to resolve this but in the process started questioning the entire KeyDict design pattern. Can you suggest what we should do about this issue?

**Comment by bqpd** [comment]
This is fixed on one of the currently open PRs.

**Comment by bqpd** [comment]
Fixed by #854
